ndarray/ndarray-core/src/commonMain/kotlin/io/kinference/ndarray/Utils.kt

```diff
@@ -106,3 +109,32 @@ internal fun IntArray.swap(leftIdx: Int, rightIdx: Int) {
     this[leftIdx] = this[rightIdx]
     this[rightIdx] = temp
 }
+/*
+ * Parallelize with batching by minDataPerLaunch
+ */
+suspend fun parallelizeByBlocks(blockSize: Int,
+                                countBlocks: Int,
+                                minDataPerLaunch: Int,
+                                function: (blockStart: Int, blockEnd: Int) -> Unit) {
+    val batchSize = run {
+        var batchSize = 1
+        while (batchSize < countBlocks && batchSize * blockSize < minDataPerLaunch) {
+            batchSize++
+        }
+        batchSize
+    }
+
+    if (batchSize == countBlocks) {
+        function(0, countBlocks)
+    } else {
+        coroutineScope {
+            for (blockStart in 0 until countBlocks step batchSize) {
+                launch {
+                    function(blockStart, min(blockStart + batchSize, countBlocks))
+                }
+            }
+        }
+    }
+}
+
+suspend inline fun parallelizeByRows(rowSize: Int, countRows: Int, minDataPerLaunch: Int, noinline function: (rowStart: Int, rowEnd: Int) -> Unit) = parallelizeByBlocks(rowSize, countRows, minDataPerLaunch, function)
```

Comment:
Why do you need suspend modifier for inline function? You can just inline internal callback as well, and then inline function will work in every code: both suspendable and non-suspendable.